### PR TITLE
Finalize multidimensional resource merge support

### LINKED ISSUE #130: Support for merging play resources
Currently plugin supports merging of flavors, but if you have a multi dimensional flavors like `stagingFree`, `stagingPaid`, `productionFree` and `productionPaid`, then folders like `src/staging/play` or `src/production/play` are ignored.

### PR DESCRIPTION
**Pull request by SUPERCILEX**
Built on top of #308 

Fixes #130

### CODE REVIEW

**Comment by ChristianBecker** [comment]
@SUPERCILEX : In this PR you changed the expected structure of the plugin. Basically
- `play/de-DE/listing/…` to `play/listings/de-DE/…`
- `play/de-DE/whatsnew` to `play/release-notes/de-DE/{track}.txt`

Can you explain shortly why the language isn't at the top-level any longer?

**Comment by SUPERCILEX** [comment]
Yeah, a few reasons:
1. I feel like that structure makes it easier to see what's what, aka it's clearer IMO
1. The in-app purchases PR doesn't use a locale so this is the only structure that works in that context
1. It lets us make assumptions easier (the code ended up being simpler that way)

**Comment by bhurling** [comment]
The code looks good. I'm just a little worried about migration. I think we can't just tell people to wipe their play folder and run the bootstrap task again. 1. They would loose their screenshots as we can only download thumbnails, 2. Any changes they might have made during development are lost too (like preparing the release notes for the next upload)

Any idea how we can make this migration a little smoother?

**Comment by SUPERCILEX** [comment]
Yeah, I had a few options in mind:
1. The bootstrap thing
1. Writing a shell script to move the files around
1. Providing an independent jar that people can run since they obviously have the JDK
1. Creating a temporary task in the plugin itself to migrate people that we would keep just for the initial v2.0 release

All of these are feasible, it's just a matter of what skills we feel we have and what would work best for people upgrading. I personally like the jar idea because it doesn't clutter the plugin but still lets us use a language we're all proficient in. The temporary task would work too.

I think once all of our development is done and we're reaching maturity, we'll have to invest a significant amount of time coming up with a migration solution everyone is happy with.

**Comment by bhurling** [comment]
shell script or jar sounds good. We could host it in our repo and provide a quick shell command that would download and run it. Similar to how you would install [homebrew](https://brew.sh/)

**Comment by bhurling** [comment]
On top of that, our `publish` tasks should check if metadata is still in the old format and either throw an error or print a warning (including a link to the migration guide)

File: plugin/src/main/kotlin/com/github/triplet/gradle/play/internal/Listings.kt [hunk 1]
```diff
@@ -3,6 +3,9 @@ package com.github.triplet.gradle.play.internal
 import java.io.File
 import java.io.FileFilter
 
+internal const val RELEASE_NOTES_DEFAULT_NAME = "default.txt"
```
[the remark this answers is not in the document]
Comment (SUPERCILEX) [reply plugin/src/main/kotlin/com/github/triplet/gradle/play/internal/Listings.kt]: I'm changing that in another PR, but yeah, let's be consistent for now.

File: plugin/src/test/groovy/com/github/triplet/gradle/play/GenerateResourcesTest.groovy [hunk 2]
```diff
@@ -147,8 +148,216 @@ class GenerateResourcesTest {
...
+    @Test
+    void multidimensionalResourcesAreNotOverwritten() {
+        def project = TestHelper.evaluatableProject()
+        def originalReleaseNotes = new File(TestHelper.FIXTURE_WORKING_DIR,
+                'src/flavor1Flavor3BuildType1/play/release-notes/en-US/default.txt').text
+        def originalFullDescription = new File(TestHelper.FIXTURE_WORKING_DIR,
+                'src/flavor1Flavor3BuildType1/play/listings/en-US/fulldescription').text
+
+        project.android {
+            flavorDimensions 'diem1', 'diem2'
```
[the remark this answers is not in the document]
Comment (SUPERCILEX) [reply plugin/src/test/groovy/com/github/triplet/gradle/play/GenerateResourcesTest.groovy]: No prob, sounds like a good idea

File: plugin/src/test/groovy/com/github/triplet/gradle/play/GenerateResourcesTest.groovy [hunk 3]
```diff
@@ -147,8 +148,216 @@ class GenerateResourcesTest {
...
+                'build/outputs/play/flavor1Flavor3BuildType1/res/release-notes/en-US/default.txt').text
+        def processedFullDescription = new File(TestHelper.FIXTURE_WORKING_DIR,
+                'build/outputs/play/flavor1Flavor3BuildType1/res/listings/en-US/fulldescription').text
+
+        assertEquals(originalReleaseNotes, processedReleaseNotes)
+        assertEquals(originalFullDescription, processedFullDescription)
+    }
+
+    @Test(expected = TaskExecutionException)
+    void conflictingFlavorsThrows() {
```
[the remark this answers is not in the document]
Comment (SUPERCILEX) [reply plugin/src/test/groovy/com/github/triplet/gradle/play/GenerateResourcesTest.groovy]: I think you might be confused, I'm throwing if the _flavors_ conflict.
```
[main, staging, free, freeStaging, dogfood, freeStagingDogfood]
```
That is, if there are two identically named resources in `staging` and `free` we don't know which one should take priority. That's different than the flavor _dimensions_. TBH, there are way too many similar terms so I might be getting confused myself. 😆

File: plugin/src/test/groovy/com/github/triplet/gradle/play/GenerateResourcesTest.groovy [hunk 4]
```diff
@@ -147,8 +148,216 @@ class GenerateResourcesTest {
...
+
+        project.tasks.clean.execute()
+        project.tasks.generateFlavor1Flavor3ReleasePlayResources.execute()
+
+        def processedFullDescription = new File(TestHelper.FIXTURE_WORKING_DIR,
+                'build/outputs/play/flavor1Flavor3Release/res/listings/en-US/fulldescription').text
+        def processedShortDescription = new File(TestHelper.FIXTURE_WORKING_DIR,
+                'build/outputs/play/flavor1Flavor3Release/res/listings/en-US/shortdescription').text
+
+        assertEquals(originalFullDescription, processedFullDescription)
```
[the remark this answers is not in the document]
Comment (SUPERCILEX) [reply plugin/src/test/groovy/com/github/triplet/gradle/play/GenerateResourcesTest.groovy]: Yup, I already had a test for that so I simply removed the full desc stuff.

File: plugin/src/test/groovy/com/github/triplet/gradle/play/GenerateResourcesTest.groovy (hunk 3 above)
[the remark this answers is not in the document]
Comment (bhurling) [reply plugin/src/test/groovy/com/github/triplet/gradle/play/GenerateResourcesTest.groovy]: Actually you have a nice example here. `free` should take precedence over `staging` as `free` is part of the dimension `pricing` which is the first one in the list of dimensions. Android's conflict resolution does the same

[the remark this answers is not in the document]
Comment (SUPERCILEX) [reply plugin/src/test/groovy/com/github/triplet/gradle/play/GenerateResourcesTest.groovy]: Ohhhhhhhhh, gotya. So the order of the dimension influences the order of the flavor. Will fix. 👍

**Pull request merged**